### PR TITLE
Add a new interface: dg interface

### PR DESCRIPTION
**Pull request by markcdg**
Here is a new interface for the DG BEACON tool from DG Technologies, Inc. (www.dgtech.com).
Note: we changed the version to 3.2.2, but you will need set it the correct version.

### CODE REVIEW

File: can/interfaces/dg/dg.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+                            | self.beacon.FILTER_FLAG_SAMPLING_INACTIVE)
+        dataFil["filter_blocks"] = []
+        for item in filters:
+            dataFil["filter_blocks"].append({})
+            dataFil["filter_blocks"][counter]["byte_offset"] = 0
+            dataFil["filter_blocks"][counter]["data_type"] = self.beacon.FILTER_DATA_TYPE_HEADER
+            dataFil["filter_blocks"][counter]["operator"] = self.beacon.BIT_FIELD_CHECK
+            dataFil["filter_blocks"][counter]["mask"] = self._int2list(item["can_mask"])
+            dataFil["filter_blocks"][counter]["pattern"] = self._int2list(item["can_id"])
+            counter += 1
```
Comment (christiansandberg) [inline can/interfaces/dg/dg.py]: What is counter used for?

File: doc/interfaces/dg.rst [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+Bus
+---
+
+.. autoclass:: can.interfaces.dg.dg.dgBus
+
+
+Configuration
+-------------
+There are 6 parameters accepted by the ``dgBus`` class::
```
Comment (christiansandberg) [inline doc/interfaces/dg.rst]: Document using Sphinx instead.

File: can/interfaces/dg/dg.py (hunk 1 above)
Comment (mikado98765) [reply can/interfaces/dg/dg.py]: The counter is used to add and index the proper filter in "filters"

Comment (christiansandberg) [reply can/interfaces/dg/dg.py]: Oh sorry, stupid me :)

File: can/interfaces/dg/dg.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+# ----------------------------------------------------------------------
+import weakref
+# import sys
+from can import BusABC, Message
+from can.interfaces.dg.dg_gryphon_protocol import server_commands
+from can.broadcastmanager import LimitedDurationCyclicSendTaskABC, ModifiableCyclicTaskABC
+
+
+class dgBus(BusABC):
```
[the remark this answers is not in the document]
Comment (markcdg) [reply can/interfaces/dg/dg.py]: Christian, Thank you for all your comments. We will resolve these issues.

File: doc/interfaces/dg.rst (hunk 2 above)
Comment (mikado98765) [reply doc/interfaces/dg.rst]: When I build the docs using Sphinx from the directions in the developers tutorial, I just get html files in doc/_build. Is there any specific style I should be adhering to when doing the rst documentation?

Comment (christiansandberg) [reply doc/interfaces/dg.rst]: Sorry, that was confusing of me. You can have a look at some other interface init method’s docstring for examples. You could also utilize the typing module to document the types accepted by the parameters but that’s not required (I think).

**Comment by mikado98765** [comment]
We are aware we missed the 'dgBus' in can/interfaces/__init__.py. We'll fix that.

File: can/interfaces/dg/dg.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (markcdg) [reply can/interfaces/dg/dg.py]: Change dgBus to DGBus. Done.

File: can/interfaces/dg/dg.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+#
+# File Name: dg.py
+# Author(s): mohtake <mohtake@dgtech.com>
+# Target Project: python-can
+# Description:
+# Notes:
+# ----------------------------------------------------------------------
+# Release Revision: $Rev: 74 $
+# Release Date: $Date: 2019/08/05 21:07:52 $
+# Revision Control Tags: $Tags: tip $
```
[the remark this answers is not in the document]
Comment (markcdg) [reply can/interfaces/dg/dg.py]: Removed. Done.

File: can/interfaces/dg/dg.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+from can.interfaces.dg.dg_gryphon_protocol import server_commands
+from can.broadcastmanager import LimitedDurationCyclicSendTaskABC, ModifiableCyclicTaskABC
+
+
+class dgBus(BusABC):
+    """ Beacon python-can backend API """
+
+    # Init method
+    def __init__(self, chan=1, mode="CAN", ip="localhost", bitrate=500000, termination=True,
+                 databitr=2000000, **kwargs):
```
[the remark this answers is not in the document]
Comment (markcdg) [reply can/interfaces/dg/dg.py]: Fixed.

File: can/interfaces/dg/dg.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        self.bitrate = dgBus._int2list(bitrate)
+        self.bitrate.reverse()
+        self.termination = [1] if termination else [0]
+        if mode == "CAN":
+            self.mode = [0]
+        else:
+            if mode == "CANFD":
+                self.mode = [1]
+            else:
+                self.mode = [2]
```
[the remark this answers is not in the document]
Comment (markcdg) [reply can/interfaces/dg/dg.py]: Fixed.

File: can/interfaces/dg/dg.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+        for item in datab:
+            if item < 16:
+                header = header + '0' + hex(item)[2:]
+            else:
+                header = header + hex(item)[2:]
+        return int(header, 16)
+
+    # Method to decode Message type to gryph dict
+    @classmethod
+    def decode_msg(cls, msg):
```
[the remark this answers is not in the document]
Comment (markcdg) [reply can/interfaces/dg/dg.py]: Changed. Done.

File: doc/interfaces/dg.rst (hunk 2 above)
Comment (markcdg) [reply doc/interfaces/dg.rst]: Done.

**Comment by zariiii9003** [comment]
Maybe it would make sense to export most of the code to an external package and import it as an optional dependency. 
Right now your code is almost as big as all the other interfaces combined. Who is going to maintain that?

**Comment by markcdg** [comment]
> Maybe it would make sense to export most of the code to an external package and import it as an optional dependency.
> Right now your code is almost as big as all the other interfaces combined. Who is going to maintain that?
You are correct. I will pull out the module for the BEACON tool into an external package.

**Comment by markcdg** [comment]
I am still having an issue getting this pull request to complete the testing. As requested by zariiii9003, I have removed the files. Should I just open a new pull request?

**Comment by zariiii9003** [comment]
The log file says you have to fix the syntax in the `setup.py`:

```
Processing /home/travis/build/hardbyte/python-can
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-req-build-x978ibe6/setup.py", line 36
        <<<<<<< HEAD
         ^
    SyntaxError: invalid syntax
```

File: can/interfaces/dg/dg.py [hunk 8]
```diff
@@ -0,0 +1,456 @@
...
+        self.beacon.CMD_CARD_IOCTL(
+            self.channel, self.beacon.IOCTL_GSETBITRATE, data_in=temp_bitr
+        )
+        self.beacon.CMD_CARD_IOCTL(
+            self.channel, self.beacon.IOCTL_GCANSETMODE, data_in=temp_mode
+        )
+        self.beacon.CMD_INIT(self.channel, value_in=self.beacon.ALWAYS_INIT)
+        self.beacon.CMD_CARD_SET_FILTER_MODE(
+            self.channel, self.beacon.FILTER_OFF_PASS_ALL
+        )
```
Comment (hardbyte) [inline can/interfaces/dg/dg.py:103]: I assume there isn't any downside to calling this, and then a few lines down via the `super` calling `set_filters`?

File: can/interfaces/dg/dg.py [hunk 9]
```diff
@@ -0,0 +1,456 @@
...
+        self.channel_info = ("dg channel '%s' on " + self.mode) % self.channel
+        super(DGBus, self).__init__(self.channel, None)
```
Comment (hardbyte) [inline can/interfaces/dg/dg.py:105]: Something seems fishy here. Can you please explicitly deal with `can_filters` 
As written I think they are just being dropped?

File: can/interfaces/dg/dg.py [hunk 10]
```diff
@@ -0,0 +1,456 @@
...
+                "data_type"
+            ] = self.beacon.FILTER_DATA_TYPE_HEADER
+            dataFil["filter_blocks"][counter]["operator"] = self.beacon.BIT_FIELD_CHECK
+            dataFil["filter_blocks"][counter]["mask"] = self._int_to_list(
+                item["can_mask"]
+            )
+            dataFil["filter_blocks"][counter]["pattern"] = self._int_to_list(
+                item["can_id"]
+            )
+            counter += 1
```
Comment (hardbyte) [inline can/interfaces/dg/dg.py:280]:
```suggestion
        for item in filters:
            dataFil["filter_blocks"].append({
                "byte_offset": 0,
                "data_type": self.beacon.FILTER_DATA_TYPE_HEADER,
                "operator": self.beacon.BIT_FIELD_CHECK,
                "mask": self._int_to_list(item["can_mask"]),
                "pattern": self._int_to_list(item["can_id"])
              })
```

File: can/interfaces/dg/dg.py [hunk 11]
```diff
@@ -0,0 +1,456 @@
...
+        self.beaconref = _beacon
+        self.channel = channel
+
+    def modify_data(self, message):
+        self.beaconref().CMD_SCHED_MSG_REPLACE(
+            self.idIn, DGBus.msg_to_dict(message), index=0
+        )
+
+    def stop(self):
+        self.beaconref().CMD_SCHED_KILL_TX(self.channel, self.idIn)
```
Comment (hardbyte) [inline can/interfaces/dg/dg.py:456]:
```suggestion
        self.beaconref().CMD_SCHED_KILL_TX(self.channel, self.idIn)

```

**Review by hardbyte — CHANGES_REQUESTED** [review]
Thanks for the PR, everything is looking pretty good. I've left comments inline. 
How did you get on splitting out a new Python library implementing the DG Gryphon Protocol?

I'd like to see tests for the utility functions `_dict_to_msg` and `_list_to_int` and their inverse.

There is still a `.pyc` file which should be removed.

File: doc/interfaces/dg.rst [hunk 12]
```diff
@@ -0,0 +1,23 @@
+DG Interface
+================
+
+Interface to `DG Technologies <http://dgtech.com>`__ Beacon
+
```
Comment (felixdivo) [inline doc/interfaces/dg.rst:5]: Are there any extra drivers required for this? What platforms (operating systems) does it run on?

File: doc/interfaces/dg.rst [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+---
+
+.. autoclass:: can.interfaces.dg.dg.DGBus
+
+
+Capability
+-------------
+The ``DGBus`` class implements several python-can methods::
+
+    send, send_periodic, recv, set_filters, flush_tx_buffer, _detect_available_configs, shutdown
```
Comment (felixdivo) [inline doc/interfaces/dg.rst:17]: I think there is no need to mention the methods with definitely need to be implemented anyways. It raises the question if other interfaces do it even if they do not mention it. I'd suggest including these:

```suggestion
The ``DGBus`` class implements several optional methods: ``send_periodic`` and ``set_filters`` within the driver, as well as ``flush_tx_buffer`` and ``_detect_available_configs``.
```

**Comment by felixdivo** [comment]
Cool mocking tests!

**Comment by felixdivo** [comment]
@markcdg Do you plan on working further on this PR? It looks like a neat addition to the library!

**Comment by felixdivo** [comment]
I opened #1132 to track this topic and will close this one as there is apparently no development effort going on.

**Pull request closed without merging**